### PR TITLE
Creating AddonQRCode component

### LINKED ISSUE #10800: Implement an AddonQRCode component
For the Mobile Add-Ons Promo project, we need a box that appears that contains a QR code, as well as some text that explains what the user is seeing. According to https://docs.google.com/document/d/1upD3G06aJFZg4XjhIeQ-MKo494ny-7geOu3YPyxXH3Q/edit, the code should appear in a black frame, and should have the following text underneath it:

"To get {addonName} on Firefox for Android, point your device camera to the code above or copy this link"

Sample image:

Screen Shot 2021-07-27 at 2 37 41 PM

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #10800 

Not try to keep this patch small, I have not included the code that will include this component on the Add-On Detail page. Here is a screenshot of what it will looks like:

Screen Shot 2021-09-16 at 11 48 58

### CODE REVIEW

File: src/amo/components/AddonQRCode/index.js [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+  componentDidMount() {
+    this.onInteract(ADDON_QRCODE_IMPRESSION_ACTION);
+  }
+
+  render(): React.Node {
+    const { addon, i18n, onDismiss } = this.props;
+
+    const idAsString = String(addon.id);
+    const imgSrc = qrCodeSrcs[idAsString];
+    invariant(imgSrc, `Could not find a QR code for addonId: ${idAsString}`);
```
Comment (bobsilverberg) [inline src/amo/components/AddonQRCode/index.js:109]: Note that, for the purposes of this temporary experiment, I am treating the source of truth for which add-ons can/should display a QR code, to be the list maintained in this component (in `qrCodeSrcs`). This allows us to have this `invariant` here, which will throw if we try to render this component with an invalid add-on. If we also use this list on the add-on detail page to determine whether or not to display a link, then we are guaranteed that this error should not happen.

File: src/amo/components/AddonQRCode/index.js [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  '722': qrCode722,
+  '735894': qrCode735894,
+  '811592': qrCode811592,
+  '824288': qrCode824288,
+  '855413': qrCode855413,
+  '866226': qrCode866226,
+  '869140': qrCode869140,
+  '953945': qrCode953945,
+  '9609': qrCode9609,
+};
```
Comment (eviljeff) [inline src/amo/components/AddonQRCode/index.js:58]: I can't help thinking there should be a more efficient way of doing this than importing each image file and then building a data structure referencing the imports but 🤷‍♂️ experiment I guess.

Comment (eviljeff) [reply src/amo/components/AddonQRCode/index.js:58]: Is just a list of the ids and loading the specific image on demand possible/practical?

File: src/amo/components/AddonQRCode/index.js [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+      utm_source: DEFAULT_UTM_SOURCE,
+      utm_medium: DEFAULT_UTM_MEDIUM,
+      utm_campaign: ADDON_QRCODE_CAMPAIGN,
+      utm_content: String(idAsString),
+    });
+    addonUrl = `/${CLIENT_APP_ANDROID}${addonUrl}`;
+
+    const labelText = i18n.sprintf(
+      i18n.gettext(
+        'To get %(addonName)s on Firefox for Android, point your device camera to the code above or copy %(linkStart)sthis link%(linkEnd)s',
```
Comment (eviljeff) [inline src/amo/components/AddonQRCode/index.js:121]: I'm aware this is from the PRD but I'm not sure this is descriptive/accurate enough - it reads to me like you should open the camera app on your phone, which won't scan the QR code - you need to choose the scan button within Firefox.

**Review by eviljeff — APPROVED** [review]

File: src/amo/components/AddonQRCode/index.js (hunk 2 above)
Comment (willdurand) [reply src/amo/components/AddonQRCode/index.js:58]: I don't really like all the imports either. I am wondering if we could use URL paths instead (and find a way to copy the files to the `dist/` folder with webpack).

Comment (bobsilverberg) [reply src/amo/components/AddonQRCode/index.js:58]: I agree it's ugly, but I'm not sure of another way to do it with the files living locally. It's possible we could do something fancy with Webpack, but is it worth it for an experiment? 

I think if the experiment is successful and we decide to move forward with this, we quite likely will end up generating these on the server and returning a URL to the image file in an API response, much like we do for other add-on images. For this reason I'm not convinced it's worth putting a lot of time into a better temporary solution. Having said that, if you have any specific suggestions of how this might work I'm happy to try.

One advantage to this approach, as I mentioned in a separate comment, is that at least we have a definitive list of all add-ons for which we can possibly display a QR code. Using this list to control the display of links limits the possibility of displaying a link to an invalid QR code, which would then either display nothing, or throw an exception. Without this safety net, we'd need extra checks to make sure we're not trying to display an invalid add-on in the panel.

I know it's all a bit hacky, but it actually works quite well, and is temporary.

File: src/amo/components/AddonQRCode/index.js (hunk 3 above)
Comment (bobsilverberg) [reply src/amo/components/AddonQRCode/index.js:121]: If you have an iPhone, simply opening the camera and pointing it at a QR code *will* scan it, at least on my phone. I imagine that's the scenario that we are looking at here. I suppose to be more specific we could say "scan the code above with your camera" or something like that? I'm not sure how we came up with the instructions for that.

File: src/amo/components/AddonQRCode/index.js (hunk 2 above)
Comment (bobsilverberg) [reply src/amo/components/AddonQRCode/index.js:58]: I'll play around a bit with Webpack to see if I can do something with it. If it's not too much work then I think it's worth it. It's just that messing with Webpack scares me. 😬

File: src/amo/components/AddonQRCode/index.js (hunk 3 above)
Comment (willdurand) [reply src/amo/components/AddonQRCode/index.js:121]: if you have an IPhone, can you install addons?

Comment (bobsilverberg) [reply src/amo/components/AddonQRCode/index.js:121]: > if you have an iPhone, can you install addons?

Um, good point! I don't have an Android phone so I guess I'm not the target audience. 😉 What is the experience like on an Android phone, I guess that's what we need to know.

File: src/amo/components/AddonQRCode/index.js (hunk 2 above)
Comment (bobsilverberg) [reply src/amo/components/AddonQRCode/index.js:58]: Will and I have been discussing this on Slack. I was able to get Webpack to copy the files to the `/dist` folder, but I'm not sure what it's going to take to make them usable both locally and when deployed. He has offered to take a look. Here's a diff of the changes I made:

```diff
diff --git a/package.json b/package.json
index ca155dbc8..e6ca2b05c 100644
--- a/package.json
+++ b/package.json
@@ -172,6 +172,7 @@
     "compression": "1.7.4",
     "config": "3.3.6",
     "connected-react-router": "5.0.1",
+    "copy-webpack-plugin": "9.0.1",
     "core-js": "3.17.3",
     "deep-eql": "4.0.0",
     "deepcopy": "2.1.0",
diff --git a/webpack-common.js b/webpack-common.js
index eaa645691..4a20e6105 100644
--- a/webpack-common.js
+++ b/webpack-common.js
@@ -1,6 +1,9 @@
 /* eslint-disable import/no-extraneous-dependencies */
+import path from 'path';
+
 import autoprefixer from 'autoprefixer';
 import CircularDependencyPlugin from 'circular-dependency-plugin';
+import CopyPlugin from 'copy-webpack-plugin';
 import config from 'config';
 import MiniCssExtractPlugin from 'mini-css-extract-plugin';
 import webpack from 'webpack';
@@ -143,6 +146,20 @@ export function getPlugins({ withBrowserWindow = true } = {}) {
       /locale$/,
       new RegExp(`^\\.\\/.*?\\/amo\\.js$`),
     ),
+    new CopyPlugin({
+      patterns: [
+        {
+          from: path.resolve(
+            __dirname,
+            'src',
+            'amo',
+            'components',
+            'AddonQRCode',
+            'img',
+          ),
+        },
+      ],
+    }),
   ];

   if (withBrowserWindow) {
```

File: src/amo/components/AddonQRCode/index.js (hunk 3 above)
Comment (eviljeff) [reply src/amo/components/AddonQRCode/index.js:121]: > I'm not sure how we came up with the instructions for that.

If I had to guess, I'd say the person who came up with instructions had an iPhone and assumed the experience was the same on other platforms 😬

The experience for me on Android (specifically for me - from a quick search the behaviour across different manufactures and models is [gloriously inconsistent](https://www.qr-code-generator.com/guides/scan-qr-code-android/)) is the camera app doesn't automatically detect the QR code - I either have to press a button in the camera app to instead open a Lens app that has all kinds of image detection built in to it, or use the QR button in Firefox itself (what I would do).  
I wonder if there's a support article already about QR support in Firefox we could link to

Comment (eviljeff) [reply src/amo/components/AddonQRCode/index.js:121]: > Um, good point! I don't have an Android phone 

-> service-now request 😉

Comment (willdurand) [reply src/amo/components/AddonQRCode/index.js:121]: On my Android phone(s), I have to install a "QR Scanner" app, which is then used directly or by other apps to scan QR codes.

File: src/amo/components/AddonQRCode/index.js (hunk 2 above)
Comment (willdurand) [reply src/amo/components/AddonQRCode/index.js:58]: On top of this PR, and thanks to your patch above, I made some changes to simplify the code a bit and remove all the imports. This works in local dev and it should work on hosted envs as well:

```diff
diff --git a/config/default.js b/config/default.js
index c2ccb1a9e..b806ef849 100644
--- a/config/default.js
+++ b/config/default.js
@@ -111,6 +111,7 @@ module.exports = {
     'mozillaUserId',
     'restrictSearchResultsToAppVersion',
     'rtlLangs',
+    'staticPath',
     'trackingEnabled',
     'trackingId',
     'trackingSendInitPageView',
diff --git a/config/development-local.js b/config/development-local.js
index e700f3ed7..a2e9f444d 100644
--- a/config/development-local.js
+++ b/config/development-local.js
@@ -1,3 +1,5 @@
+import { staticPath } from './lib/shared';
+
 // This config should be used with a local addons-server setup.
 module.exports = {
   apiHost: 'http://olympia.test',
@@ -11,6 +13,10 @@ module.exports = {
 
   mozillaUserId: 10968,
   CSP: false,
+
+  // `config/development.js` overrides the static path to use
+  // webpack-dev-server but we don't want that for this environment.
+  staticPath,
   enableNodeStatics: true,
 
   // See: https://github.com/mozilla/addons-frontend/issues/10545
diff --git a/config/development.js b/config/development.js
index 9c8a0c269..1165749c4 100644
--- a/config/development.js
+++ b/config/development.js
@@ -3,7 +3,7 @@ import { addonsServerDevCDN, analyticsHost, apiDevHost } from './lib/shared';
 
 const webpackServerHost = process.env.WEBPACK_SERVER_HOST || '127.0.0.1';
 const webpackServerPort = 3001;
-const webpackHost = `${webpackServerHost}:${webpackServerPort}`;
+const webpackDevServer = `${webpackServerHost}:${webpackServerPort}`;
 
 module.exports = {
   apiHost: 'http://localhost:3000',
@@ -38,7 +38,10 @@ module.exports = {
   serverPort: 3333,
   webpackServerHost,
   webpackServerPort,
-  webpackHost,
+
+  // In local dev, we serve static files using webpack-dev-server.
+  // We need to remove the protocol because of `yarn amo:dev-https`.
+  staticPath: `//${webpackDevServer}/`,
 
   CSP: {
     directives: {
@@ -47,25 +50,25 @@ module.exports = {
         addonsServerDevCDN,
         analyticsHost,
         apiDevHost,
-        webpackHost,
+        webpackDevServer,
         // This is needed for pino-devtools.
         `${webpackServerHost}:3010`,
       ],
       fontSrc: [
-        webpackHost,
+        webpackDevServer,
       ],
       imgSrc: [
         "'self'",
         'data:',
         addonsServerDevCDN,
-        webpackHost,
+        webpackDevServer,
       ],
       scriptSrc: [
         "'self'",
         // webpack injects inline JS
         "'unsafe-inline'",
         addonsServerDevCDN,
-        webpackHost,
+        webpackDevServer,
         `${analyticsHost}/analytics.js`,
       ],
       styleSrc: [
diff --git a/package.json b/package.json
index ca155dbc8..e6ca2b05c 100644
--- a/package.json
+++ b/package.json
@@ -172,6 +172,7 @@
     "compression": "1.7.4",
     "config": "3.3.6",
     "connected-react-router": "5.0.1",
+    "copy-webpack-plugin": "9.0.1",
     "core-js": "3.17.3",
     "deep-eql": "4.0.0",
     "deepcopy": "2.1.0",
diff --git a/src/amo/components/AddonQRCode/index.js b/src/amo/components/AddonQRCode/index.js
index e2af690b4..851194510 100644
--- a/src/amo/components/AddonQRCode/index.js
+++ b/src/amo/components/AddonQRCode/index.js
@@ -2,6 +2,7 @@
 import invariant from 'invariant';
 import * as React from 'react';
 import { compose } from 'redux';
+import config from 'config';
 
 import {
   CLIENT_APP_ANDROID,
@@ -21,41 +22,12 @@ import type { ElementEvent } from 'amo/types/dom';
 import type { I18nType } from 'amo/types/i18n';
 
 import './styles.scss';
-import qrCode229918 from './img/229918.png';
-import qrCode328839 from './img/328839.png';
-import qrCode506646 from './img/506646.png';
-import qrCode520576 from './img/520576.png';
-import qrCode521554 from './img/521554.png';
-import qrCode607454 from './img/607454.png';
-import qrCode627490 from './img/627490.png';
-import qrCode722 from './img/722.png';
-import qrCode735894 from './img/735894.png';
-import qrCode811592 from './img/811592.png';
-import qrCode824288 from './img/824288.png';
-import qrCode855413 from './img/855413.png';
-import qrCode866226 from './img/866226.png';
-import qrCode869140 from './img/869140.png';
-import qrCode953945 from './img/953945.png';
-import qrCode9609 from './img/9609.png';
-
-export const qrCodeSrcs = {
-  '229918': qrCode229918,
-  '328839': qrCode328839,
-  '506646': qrCode506646,
-  '520576': qrCode520576,
-  '521554': qrCode521554,
-  '607454': qrCode607454,
-  '627490': qrCode627490,
-  '722': qrCode722,
-  '735894': qrCode735894,
-  '811592': qrCode811592,
-  '824288': qrCode824288,
-  '855413': qrCode855413,
-  '866226': qrCode866226,
-  '869140': qrCode869140,
-  '953945': qrCode953945,
-  '9609': qrCode9609,
-};
+
+// TODO: move to config
+const EXPECTED_ADDON_IDS = [
+  229918, 328839, 506646, 520576, 521554, 607454, 627490, 722, 735894, 811592,
+  824288, 855413, 866226, 869140, 953945, 9609,
+];
 
 export const ADDON_QRCODE_CAMPAIGN = 'addon-qr-code';
 export const ADDON_QRCODE_CATEGORY = 'Addon QR Code';
@@ -63,7 +35,7 @@ export const ADDON_QRCODE_CLICK_ACTION = 'addon-qr-code-click';
 export const ADDON_QRCODE_IMPRESSION_ACTION = 'addon-qr-code-impression';
 
 type Props = {|
-  addon: AddonType,
+  addon?: AddonType,
   onDismiss: (e: ElementEvent | null) => void,
 |};
 
@@ -104,15 +76,20 @@ export class AddonQRCodeBase extends React.Component<InternalProps> {
   render(): React.Node {
     const { addon, i18n, onDismiss } = this.props;
 
-    const idAsString = String(addon.id);
-    const imgSrc = qrCodeSrcs[idAsString];
-    invariant(imgSrc, `Could not find a QR code for addonId: ${idAsString}`);
+    if (!addon) {
+      return null;
+    }
+
+    invariant(
+      EXPECTED_ADDON_IDS.includes(addon.id),
+      'add-on should have a QR code',
+    );
 
     let addonUrl = addQueryParams(getAddonURL(addon.slug), {
       utm_source: DEFAULT_UTM_SOURCE,
       utm_medium: DEFAULT_UTM_MEDIUM,
       utm_campaign: ADDON_QRCODE_CAMPAIGN,
-      utm_content: String(idAsString),
+      utm_content: String(addon.id),
     });
     addonUrl = `/${CLIENT_APP_ANDROID}${addonUrl}`;
 
@@ -169,7 +146,7 @@ export class AddonQRCodeBase extends React.Component<InternalProps> {
         <img
           alt={altText}
           className="AddonQRCode-img"
-          src={qrCodeSrcs[idAsString]}
+          src={`${config.get('staticPath')}${addon.id}.png?v=1`}
         />
         <div className="AddonQRCode-label">{labelWithLink}</div>
       </div>
diff --git a/webpack-common.js b/webpack-common.js
index eaa645691..4a20e6105 100644
--- a/webpack-common.js
+++ b/webpack-common.js
@@ -1,6 +1,9 @@
 /* eslint-disable import/no-extraneous-dependencies */
+import path from 'path';
+
 import autoprefixer from 'autoprefixer';
 import CircularDependencyPlugin from 'circular-dependency-plugin';
+import CopyPlugin from 'copy-webpack-plugin';
 import config from 'config';
 import MiniCssExtractPlugin from 'mini-css-extract-plugin';
 import webpack from 'webpack';
@@ -143,6 +146,20 @@ export function getPlugins({ withBrowserWindow = true } = {}) {
       /locale$/,
       new RegExp(`^\\.\\/.*?\\/amo\\.js$`),
     ),
+    new CopyPlugin({
+      patterns: [
+        {
+          from: path.resolve(
+            __dirname,
+            'src',
+            'amo',
+            'components',
+            'AddonQRCode',
+            'img',
+          ),
+        },
+      ],
+    }),
   ];
 
   if (withBrowserWindow) {
diff --git a/webpack.dev.config.babel.js b/webpack.dev.config.babel.js
index 52a484820..700489966 100644
--- a/webpack.dev.config.babel.js
+++ b/webpack.dev.config.babel.js
@@ -30,10 +30,9 @@ export const babelOptions = {
     : babelPlugins,
 };
 
-const webpackHost = config.get('webpackServerHost');
-const webpackPort = config.get('webpackServerPort');
-
-const hmr = `webpack-hot-middleware/client?path=//${webpackHost}:${webpackPort}/__webpack_hmr`;
+const hmr = `webpack-hot-middleware/client?path=${config.get(
+  'staticPath',
+)}__webpack_hmr`;
 
 const entryPoints = { [WEBPACK_ENTRYPOINT]: [hmr, 'amo/client'] };
 
@@ -51,7 +50,7 @@ export default {
     filename: '[name]-[contenthash].js',
     chunkFilename: '[name]-[contenthash].js',
     // We need to remove the protocol because of `yarn amo:dev-https`.
-    publicPath: `//${webpackHost}:${webpackPort}/`,
+    publicPath: config.get('staticPath'),
   },
   module: {
     rules: getRules({ babelOptions, bundleStylesWithJs: true }),
diff --git a/yarn.lock b/yarn.lock
index 20dcd4650..edbe67cd5 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -3660,6 +3660,19 @@ copy-descriptor@^0.1.0:
   resolved "https://registry.yarnpkg.com/copy-descriptor/-/copy-descriptor-0.1.1.tgz#676f6eb3c39997c2ee1ac3a924fd6124748f578d"
   integrity sha1-Z29us8OZl8LuGsOpJP1hJHSPV40=
 
+copy-webpack-plugin@9.0.1:
+  version "9.0.1"
+  resolved "https://registry.yarnpkg.com/copy-webpack-plugin/-/copy-webpack-plugin-9.0.1.tgz#b71d21991599f61a4ee00ba79087b8ba279bbb59"
+  integrity sha512-14gHKKdYIxF84jCEgPgYXCPpldbwpxxLbCmA7LReY7gvbaT555DgeBWBgBZM116tv/fO6RRJrsivBqRyRlukhw==
+  dependencies:
+    fast-glob "^3.2.5"
+    glob-parent "^6.0.0"
+    globby "^11.0.3"
+    normalize-path "^3.0.0"
+    p-limit "^3.1.0"
+    schema-utils "^3.0.0"
+    serialize-javascript "^6.0.0"
+
 core-js-compat@^3.9.0:
   version "3.9.1"
   resolved "https://registry.yarnpkg.com/core-js-compat/-/core-js-compat-3.9.1.tgz#4e572acfe90aff69d76d8c37759d21a5c59bb455"
@@ -5618,6 +5631,13 @@ glob-parent@^5.0.0, glob-parent@^5.1.0:
   dependencies:
     is-glob "^4.0.1"
 
+glob-parent@^6.0.0:
+  version "6.0.1"
+  resolved "https://registry.yarnpkg.com/glob-parent/-/glob-parent-6.0.1.tgz#42054f685eb6a44e7a7d189a96efa40a54971aa7"
+  integrity sha512-kEVjS71mQazDBHKcsq4E9u/vUzaLcw1A8EtUeydawvIWQCJM0qQ08G1H7/XTjFUulla6XQiDOG6MXSaG0HDKog==
+  dependencies:
+    is-glob "^4.0.1"
+
 glob-parent@~5.1.2:
   version "5.1.2"
   resolved "https://registry.yarnpkg.com/glob-parent/-/glob-parent-5.1.2.tgz#869832c58034fe68a4093c17dc15e8340d8401c4"
@@ -11052,7 +11072,7 @@ send@^0.16.1:
     range-parser "~1.2.0"
     statuses "~1.4.0"
 
-serialize-javascript@6.0.0:
+serialize-javascript@6.0.0, serialize-javascript@^6.0.0:
   version "6.0.0"
   resolved "https://registry.yarnpkg.com/serialize-javascript/-/serialize-javascript-6.0.0.tgz#efae5d88f45d7924141da8b5c3a7a7e663fefeb8"
   integrity sha512-Qr3TosvguFt8ePWqsvRfrKyQXIiW+nGbYpy8XK24NQHE83caxWt+mIymTT19DGFbNWNLfEwsrkSmN64lVWB9ag==
```

File: src/amo/components/AddonQRCode/index.js (hunk 3 above)
Comment (eviljeff) [reply src/amo/components/AddonQRCode/index.js:121]: > On my Android phone(s), I have to install a "QR Scanner" app, which is then used directly or by other apps to scan QR codes.

For you, Firefox doesn't have any support for scanning QR codes without manually installing a 3rd party app?

Comment (willdurand) [reply src/amo/components/AddonQRCode/index.js:121]: ha, I dunno. I wasn't aware of this feature in Fenix before today.

Comment (bobsilverberg) [reply src/amo/components/AddonQRCode/index.js:121]: Because the experience can be so different on different devices, and with different apps, perhaps it should just say "scan the code above with your phone"?

Comment (willdurand) [reply src/amo/components/AddonQRCode/index.js:121]: > Because the experience can be so different on different devices, and with different apps, perhaps it should just say "scan the code above with your phone"?

+1

File: src/amo/components/AddonQRCode/index.js (hunk 2 above)
Comment (bobsilverberg) [reply src/amo/components/AddonQRCode/index.js:58]: Thanks for this @willdurand! I added the code locally and it does indeed work. I guess we can test it on dev after this and the next PR are approved and merged. I'm going to make a few more changes to this PR and re-submit for review.

Comment (bobsilverberg) [reply src/amo/components/AddonQRCode/index.js]: @willdurand There is a new patch with these changes in it available for review.

File: src/amo/components/AddonQRCode/index.js (hunk 3 above)
Comment (bobsilverberg) [reply src/amo/components/AddonQRCode/index.js:90]: Meridel has agreed to take a look at this for us and provide an opinion on what copy we should use for this. I think we can land this as is for now, and I will open a follow-up to discuss changing the copy.

Comment (bobsilverberg) [reply src/amo/components/AddonQRCode/index.js:90]: I opened https://github.com/mozilla/addons-frontend/issues/10910 as the follow-up.

File: config/default.js [hunk 4]
```diff
@@ -421,4 +423,9 @@ module.exports = {
     // e.g., 20210531_amo_download_funnel_experiment: true,
     '20210714_amo_vpn_promo': true,
   },
+
+  addonIdsWithQRCodes: [
```
Comment (willdurand) [inline config/default.js:427]: I think (and I am not saying we should add that here) that we need to verify this list of IDs against the list of IDs retrieved in the `bin/fenix-qr-codes.js` script. If we can load the config in this script, that could be as easy as:

```js
const config = require('config');

// Somewhere after `Promise.all()` I suppose...

const knownIds = new Set(config.get('addonIdsWithQRCodes'));
const addonIdsFromAPI = [...new Set(addons.map(({ id }) => id))];

if (addonIdsFromAPI.filter(x => knownIds.has(x)).length > 0) {
  console.log(`'addonIdsWithQRCodes' might be outdated:`);
  console.log(`  addonIdsWithQRCodes=${JSON.stringify(knownIds.sort())}`);
  console.log(`  IDs from API=${JSON.stringify(addonIdsFromAPI)}`);
  process.exit(1);
}
```

File: bin/fenix-qr-codes.js [hunk 5]
```diff
@@ -31,7 +31,7 @@ if (!fs.existsSync(destDir)) {
         `utm_content=${id}`,
         `utm_medium=referral&utm_source=addons.mozilla.org`,
       ].join('&')
-      return QRCode.toFile(filePath, content);
+      return QRCode.toFile(filePath, content, {margin: 1});
```
Comment (willdurand) [inline bin/fenix-qr-codes.js:34]: Why is this needed?

**Review by willdurand — APPROVED** [review]
r+wc

(maybe we could `pngcrush` the images but maybe not)

File: bin/fenix-qr-codes.js (hunk 5 above)
Comment (bobsilverberg) [reply bin/fenix-qr-codes.js:34]: I was just trying to match the image from the spec as closely as possible. Without setting `margin: 1` we end up with a margin that is quite a bit wider than the image in the spec, creating a bunch of extra whitespace inside the border.

Comment (bobsilverberg) [reply bin/fenix-qr-codes.js:34]: Because this PR doesn't include any code to actually display the component, let's punt on this for now, and we can make changes to the appearance of the panel in the follow-up PR.

File: config/default.js (hunk 4 above)
Comment (bobsilverberg) [reply config/default.js]: I had to do some fiddling to deal with the `Cannot use import statement outside a module` error I was getting, but I was able to make it work. There seemed to be some kind of logic problem with your `filter`, and in trying to make it work I just removed all the `Set`s and just used arrays and it seems fine. Even if we end up with duplicate id's (which we shouldn't) it won't make any difference. I've added a new commit with my version of this change.

File: bin/fenix-qr-codes.js [hunk 6]
```diff
@@ -31,8 +33,21 @@ if (!fs.existsSync(destDir)) {
...
     })
   );
+
+  const knownIds = addonIdsWithQRCodes.addonIds;
+  const addonIdsFromAPI = addons.map(({ id }) => id);
+
+  for (const addonId of addonIdsFromAPI) {
+    if (!knownIds.includes(addonId)) {
```
Comment (willdurand) [inline bin/fenix-qr-codes.js:44]: This isn't what I wrote in the comment. I was performing a symmetric difference on purpose: we want the two lists to be the same.

Comment (willdurand) [reply bin/fenix-qr-codes.js:44]: I should have tested my code, though :p I don't think the snippet in the comment works. Let me fix it in a new PR.

Comment (bobsilverberg) [reply bin/fenix-qr-codes.js:44]: Hmm, I was thinking we just needed to check that we have a QR code for every add-on returned from the API, but I guess we also want to make sure our internal list doesn't include any addonIds that are *not* returned by the API. I think I can update my code to do that. I'm not sure we need sets. Couldn't we just sort the two arrays and do a `deepEqual` on them?
